src/renderer/atlas/AtlasEngine.cpp

```diff
@@ -315,6 +315,15 @@ CATCH_RETURN()
         }
 
         _api.selectionSpans = til::point_span_subspan_within_rect(info.selectionSpans, dr);
+
+        u32 newSelectionColor{ static_cast<COLORREF>(info.selectionBackground) | 0xff000000 };
+        if (_api.s->misc->selectionColor != newSelectionColor)
+        {
+            auto misc = _api.s.write()->misc.write();

```

Comment:
Copying from `_api.s` to `_p.s` happens in `StartPaint()`. Similar to `AtlasEngine::PaintCursor` you need to update both structs simultaneously to keep their generation count in sync.

You can test that this by setting a breakpoint into `AtlasEngine::_handleSettingsUpdate`. Then and emit an OSC 17. If done correctly, `AtlasEngine::_handleSettingsUpdate` should not get called.

Reply:
Oh no, this is interesting. It didn't cause a problem because I didn't understand your architecture.

I put the selection colors on `_api`, and I read them out of `_api`. I didn't even know (remember?) that they would also be on `_p`.

Should PaintBufferLineHelper be reading them out of `_p` instead?

Reply:
I made the call to read them out of `_p`, because that's the RenderingPayload.